### ISSUE TITLE
Using custom template, how can i call function that the tree component is on?

### ISSUE THREAD

**Issue by Gillardo** [500tech/angular2-tree-component#15]
I have a tree that displays nodes.  On each node i have a custom template, which offers the options Edit, Copy and Delete.  When one of these options is pressed, i would like to open a modal window (using ng2-bootstrap) and ask the user to confirm their action.

Now i can do this, if I create a modal window in the custom template, but that seems like a lot of HTML to repeat for each node, so instead i would like to have my ModalComponent on the same component as my tree, and somehow raise an event when a button in a template is pressed....

How can i do this?  Here is what i have so far

pages.component.html
```
<Tree #tree [nodes]="nodes" [options]="options"></Tree>

<div *ngIf="tree.activeNode">
   <!-- got modal here instead of within each node template!! //-->
    <page-copy-modal #copyModal="copyModal" [page]="tree.activeNode.data"></page-copy-modal>
</div>
```

pages.component.ts
```
import { Component, OnInit } from '@angular/core';
import { TreeComponent } from 'angular2-tree-component';
import { PagesTreeNodeComponent } from './pages-tree-node.component';
import { PageCopyModalComponent } from './page-copy-modal.component';

@Component({
    selector: 'pages',
    templateUrl: 'app/admin/page/pages.component.html',
    directives: [ TreeComponent, RouterLink, PageCopyModalComponent ]
})

export class PagesComponent implements OnInit {

    nodes: Page[];
    options: Object = {
        treeNodeTemplate: PagesTreeNodeComponent
    };

    ngOnInit() {
        // nodes loaded here
    }

   // need these functions to be called from pages-tree-node-component.ts
   onEdit () { }
   onCopy () { }
   onDelete () { }
}
```

pages-tree-node.component.html
```
{{ node.displayField }}
<div class="pull-sm-right options">
    <a (click)="onEdit()">Edit</a> |
    <a (click)="onCopy()">Copy</a> |
    <a (click)="onDelete()">Delete</a>
</div>
```

pages-tree-node.component.ts
```
import { Component, Input } from '@angular/core';
import { Router } from '@angular/router-deprecated';
import { TreeNode } from 'angular2-tree-component';

@Component({
    selector: 'pages-tree-node',
    templateUrl: 'app/admin/page/pages-tree-node.component.html'
})

export class PagesTreeNodeComponent {

    @Input()
    node: TreeNode;

    // on editing a page
    onEdit() {
        // call event on PagesComponent for edit??
    }

    onCopy() {
        // call event on PagesComponent for copy??
    }

    onDelete() {
        // call event on PagesComponent for delete??
    }
}
```

page-copy-modal.component.html
```
<div bsModal #lgModal="bs-modal" class="modal fade" tabindex="-1" role="dialog" aria-labelledby="myLargeModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title">Copy page</h4>
            </div>
            <div class="modal-body">
                Are you sure you want to copy this page, all its contents and all of it's child pages?
            </div>
            <div class="modal-footer">
                <button type="button" class="btn btn-outline-primary" (click)="onCopy()" aria-label="Copy">
                    Copy
                </button>
                <button type="button" class="btn btn-outline-secondary" (click)="lgModal.hide()" aria-label="Cancel">
                    Cancel
                </button>
            </div>
        </div>
    </div>
</div>
```

page-copy-modal.component.ts
```
import { Component, Input, ViewChild} from '@angular/core'
import { CORE_DIRECTIVES } from '@angular/common'
import { MODAL_DIRECTIVES } from 'ng2-bootstrap/components/modal';
import { ComponentsHelper } from 'ng2-bootstrap/components/utils/components-helper.service';

@Component({
    selector: 'page-copy-modal',
    directives: [ MODAL_DIRECTIVES, CORE_DIRECTIVES ],
    viewProviders: [ ComponentsHelper ],
    templateUrl: 'app/admin/page/page-copy-modal.component.html',
    exportAs: 'copyModal'
})

export class PageCopyModalComponent {

    @ViewChild('lgModal') 
    lgModal;

    @Input: any
    page;

    show(){
        this.lgModal.show();
    }

    onCopy() {
        
    }
}
```

**Comment by adamkleingit** [comment]
This is an interesting question. Sending events between components is bad practice IMO, since it's not a unidirectional data flow (if anything, rather use Redux).

I can allow passing a context to the tree node template. Something like:
```
options = {
  context: this,
  treeNodeTemplate: PagesTreeNodeComponent
}

{{ node.displayField }}
<div class="pull-sm-right options">
    <a (click)="context.onEdit()">Edit</a> |
    <a (click)="context.onCopy()">Copy</a> |
    <a (click)="context.onDelete()">Delete</a>
</div>
```

Does that sound reasonable?

**Comment by Gillardo** [comment]
Yes that could work!  Is this functionality already within the code or do you need to do work to support this, and if so, how long? :)

**Comment by Gillardo** [comment]
I agree that events is not the way forward, but this is a very good solution and i think other users will want this sort of functionality in custom templates

**Comment by 500tech-user** [comment]
It's still not in the code, I'll add it.
Hopefully I can get to it soon. Feel free to fork it and add it in the
meantime


On Thu, Jul 14, 2016 at 11:30 AM, Gillardo <notifications@github.com> wrote:

> I agree that events is not the way forward, but this is a very good
> solution and i think other users will want this sort of functionality in
> custom templates
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/500tech/angular2-tree-component/issues/15#issuecomment-232600719>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACUsQxFYkM_h-EV3cjOucPJezbHHDyCBks5qVfOIgaJpZM4JKjhd>
> .
>



-- 
*Adam Klein*
*CTO & Co-Founder*
*Tel.* +972-52-747-5633
*blog.500tech.com <http://blog.500tech.com>*
meetup.com/angularjs-il

**Comment by adamkleingit** [comment]
Published 1.1.10 to NPM which adds context to options
